Mutex groups

Bolodya1997 opened the pull request:
# Issue
There is a need for `timeout`, `refresh` chain elements to perform serial execution of { Request, Close } events for the single connection - existing sync primitives don't provide true FIFO by-ID execution.
# Solution
~`serializer.Serializer{}` - is a true FIFO by-ID lock based on channels.~
`fifosync` provides true FIFO synchronization primitives.

denis-tingajkin commented on pkg/tools/serializer/serializer.go:
```diff
@@ -0,0 +1,71 @@
+// 2. B -> lock 'a' -- B waiting 'a'
+// 3. C -> lock 'b' -- C locked 'b'
+// 4. A -> unlock 'a', D -> lock 'a' -- B locked 'a', D waiting 'a'
+type Serializer struct {
```
Could you please add benchmark tests and share the result?

Bolodya1997 replied:
Here are the benchmark tests for the `MutexGroup`:
```
Benchmark_MutexGroup_Different
Benchmark_MutexGroup_Different-4             	 3013191	       455 ns/op
Benchmark_MutexGroup_Same
Benchmark_MutexGroup_Same-4                  	 2828256	       381 ns/op
Benchmark_Executor
Benchmark_Executor-4                         	 1269505	      1019 ns/op
Benchmark_MutexGroup_Different_Goroutine
Benchmark_MutexGroup_Different_Goroutine-4   	  494902	      2116 ns/op
Benchmark_MutexGroup_Same_Goroutine
Benchmark_MutexGroup_Same_Goroutine-4        	 1000000	      3666 ns/op
Benchmark_Executor_Goroutine
Benchmark_Executor_Goroutine-4               	 1000000	      1438 ns/op
```

denis-tingajkin reviewed (commented):
Does this issue solves by private map[string]sync.Locker for `refresh`, `timeout`?

Bolodya1997 commented:
Nope, `sync.Locker` doesn't guarantee FIFO order.

denis-tingajkin commented:
@Bolodya1997 Note: `sync.Locker` is just interface... We could implement FIFOMutext that implements this interface :)

edwarnicke commented:
Or you could just use a sync.Map of serialize.Executors, which give you precisely the in-order guarantee you are looking for, and Async execution to boot.

Bolodya1997 commented:
@edwarnicke 
It was my first thought, but actually it is very bad (totally) in performance.

edwarnicke commented:
> @edwarnicke
> It was my first thought, but actually it is very bad (totally) in performance.

What were you seeing?  How were you benchmarking?

Bolodya1997 commented:
Map + bench tests:
```go
type Map struct {
	init      sync.Once
	executor  serialize.Executor
	executors map[string]*serialize.Executor
}

func (m *Map) AsyncExec(id string, f func()) (ch <-chan struct{}) {
	m.init.Do(func() {
		m.executors = map[string]*serialize.Executor{}
	})

	<-m.executor.AsyncExec(func() {
		executor, ok := m.executors[id]
		if !ok {
			executor = &serialize.Executor{}
			m.executors[id] = executor
		}
		ch = executor.AsyncExec(f)
	})
	return ch
}

func BenchmarkMutexGroup_Lock_Same(b *testing.B) {
	var g fifosync.MutexGroup
	for i := 0; i < b.N-1; i++ {
		g.Lock("a")
		g.Unlock("a")
	}
}

func BenchmarkMap_AsyncExec_Same(b *testing.B) {
	var m Map
	for i := 0; i < b.N-1; i++ {
		m.AsyncExec("a", func() {})
	}
}

func BenchmarkMutexGroup_Lock_Different(b *testing.B) {
	var g fifosync.MutexGroup
	for i := 0; i < b.N-1; i++ {
		id := uuid.New().String()
		g.Lock(id)
		g.Unlock(id)
	}
}

func BenchmarkMap_AsyncExec_Different(b *testing.B) {
	var m Map
	for i := 0; i < b.N-1; i++ {
		id := uuid.New().String()
		m.AsyncExec(id, func() {})
	}
}

func BenchmarkMutexGroup_Lock_Same_Goroutine(b *testing.B) {
	var g fifosync.MutexGroup
	for i := 0; i < b.N-1; i++ {
		go func() {
			g.Lock("a")
			g.Unlock("a")
		}()
	}
}

func BenchmarkMap_AsyncExec_Same_Goroutine(b *testing.B) {
	var m Map
	for i := 0; i < b.N-1; i++ {
		go func() {
			m.AsyncExec("a", func() {})
		}()
	}
}

func BenchmarkMutexGroup_Lock_Different_Goroutine(b *testing.B) {
	var g fifosync.MutexGroup
	for i := 0; i < b.N-1; i++ {
		go func() {
			id := uuid.New().String()
			g.Lock(id)
			g.Unlock(id)
		}()
	}
}

func BenchmarkMap_AsyncExec_Different_Goroutine(b *testing.B) {
	var m Map
	for i := 0; i < b.N-1; i++ {
		go func() {
			id := uuid.New().String()
			m.AsyncExec(id, func() {})
		}()
	}
}
```
Results:
```
BenchmarkMutexGroup_Lock_Same
BenchmarkMutexGroup_Lock_Same-4                  	 3211293	       397 ns/op
BenchmarkMap_AsyncExec_Same
BenchmarkMap_AsyncExec_Same-4                    	  728451	      1499 ns/op
BenchmarkMutexGroup_Lock_Different
BenchmarkMutexGroup_Lock_Different-4             	 1978754	       609 ns/op
BenchmarkMap_AsyncExec_Different
BenchmarkMap_AsyncExec_Different-4               	  292650	      4535 ns/op
BenchmarkMutexGroup_Lock_Same_Goroutine
BenchmarkMutexGroup_Lock_Same_Goroutine-4        	  735339	      1609 ns/op
BenchmarkMap_AsyncExec_Same_Goroutine
BenchmarkMap_AsyncExec_Same_Goroutine-4          	 1000000	      2319 ns/op
BenchmarkMutexGroup_Lock_Different_Goroutine
BenchmarkMutexGroup_Lock_Different_Goroutine-4   	 1000000	      3647 ns/op
BenchmarkMap_AsyncExec_Different_Goroutine
BenchmarkMap_AsyncExec_Different_Goroutine-4     	 1000000	      7381 ns/op
```

denis-tingajkin commented:
@Bolodya1997 please note: testing.B have functions for stopping/resuming internal timer. Please use this for preparation sections in your tests and update the results.

Bolodya1997 commented:
@denis-tingajkin 
go benchmarks setup timer for each bench test separately
```go
// runN runs a single benchmark for the specified number of iterations.
func (b *B) runN(n int) {
	benchmarkLock.Lock()
	defer benchmarkLock.Unlock()
	defer b.runCleanup(normalPanic)
	// Try to get a comparable environment for each run
	// by clearing garbage from previous runs.
	runtime.GC()
	b.raceErrors = -race.Errors()
	b.N = n
	b.parallelism = 1
	b.ResetTimer()
	b.StartTimer()
	b.benchFunc(b) // <-- test
	b.StopTimer()
	b.previousN = n
	b.previousDuration = b.duration
	b.raceErrors += race.Errors()
	if b.raceErrors > 0 {
		b.Errorf("race detected during execution of benchmark")
	}
}
```

denis-tingajkin commented:
@Bolodya1997 please take a look at official documentation https://github.com/golang/go/blob/master/src/testing/testing.go#L52-L62.

Bolodya1997 commented:
@denis-tingajkin 
Guess that starting and stopping timer around every `uuid.New().String()` can provide some problems, especially in parallel case. Replaced it with `ids` slice.
To make tests more valid, I have added cleanup of finished executors to `Map`. Here are the new results:
```
BenchmarkMutexGroup_Lock_Same
BenchmarkMutexGroup_Lock_Same-4                  	 2916673	       423 ns/op
BenchmarkMap_AsyncExec_Same
BenchmarkMap_AsyncExec_Same-4                    	  570249	      2136 ns/op
BenchmarkMutexGroup_Lock_Different
BenchmarkMutexGroup_Lock_Different-4             	 2985600	       397 ns/op
BenchmarkMap_AsyncExec_Different
BenchmarkMap_AsyncExec_Different-4               	  381249	      3244 ns/op
BenchmarkMutexGroup_Lock_Same_Goroutine
BenchmarkMutexGroup_Lock_Same_Goroutine-4        	 2210337	       505 ns/op
BenchmarkMap_AsyncExec_Same_Goroutine
BenchmarkMap_AsyncExec_Same_Goroutine-4          	  679833	      1585 ns/op
BenchmarkMutexGroup_Lock_Different_Goroutine
BenchmarkMutexGroup_Lock_Different_Goroutine-4   	  757113	      1557 ns/op
BenchmarkMap_AsyncExec_Different_Goroutine
BenchmarkMap_AsyncExec_Different_Goroutine-4     	  788718	      1692 ns/op
```
For the parallel case with different IDs the results are almost the same, but `MutexGroup` is still much more faster for all the other cases.

edwarnicke commented:
> Map + bench tests:
> 
> ```go
> type Map struct {
> 	init      sync.Once
> 	executor  serialize.Executor
> 	executors map[string]*serialize.Executor
> }
> 
> func (m *Map) AsyncExec(id string, f func()) (ch <-chan struct{}) {
> 	m.init.Do(func() {
> 		m.executors = map[string]*serialize.Executor{}
> 	})
> 
> 	<-m.executor.AsyncExec(func() {
> 		executor, ok := m.executors[id]
> 		if !ok {
> 			executor = &serialize.Executor{}
> 			m.executors[id] = executor
> 		}
> 		ch = executor.AsyncExec(f)
> 	})
> 	return ch
> }
> 
> func BenchmarkMutexGroup_Lock_Same(b *testing.B) {
> 	var g fifosync.MutexGroup
> 	for i := 0; i < b.N-1; i++ {
> 		g.Lock("a")
> 		g.Unlock("a")
> 	}
> }
> 
> func BenchmarkMap_AsyncExec_Same(b *testing.B) {
> 	var m Map
> 	for i := 0; i < b.N-1; i++ {
> 		m.AsyncExec("a", func() {})
> 	}
> }
> 
> func BenchmarkMutexGroup_Lock_Different(b *testing.B) {
> 	var g fifosync.MutexGroup
> 	for i := 0; i < b.N-1; i++ {
> 		id := uuid.New().String()
> 		g.Lock(id)
> 		g.Unlock(id)
> 	}
> }
> 
> func BenchmarkMap_AsyncExec_Different(b *testing.B) {
> 	var m Map
> 	for i := 0; i < b.N-1; i++ {
> 		id := uuid.New().String()
> 		m.AsyncExec(id, func() {})
> 	}
> }
> 
> func BenchmarkMutexGroup_Lock_Same_Goroutine(b *testing.B) {
> 	var g fifosync.MutexGroup
> 	for i := 0; i < b.N-1; i++ {
> 		go func() {
> 			g.Lock("a")
> 			g.Unlock("a")
> 		}()
> 	}
> }
> 
> func BenchmarkMap_AsyncExec_Same_Goroutine(b *testing.B) {
> 	var m Map
> 	for i := 0; i < b.N-1; i++ {
> 		go func() {
> 			m.AsyncExec("a", func() {})
> 		}()
> 	}
> }
> 
> func BenchmarkMutexGroup_Lock_Different_Goroutine(b *testing.B) {
> 	var g fifosync.MutexGroup
> 	for i := 0; i < b.N-1; i++ {
> 		go func() {
> 			id := uuid.New().String()
> 			g.Lock(id)
> 			g.Unlock(id)
> 		}()
> 	}
> }
> 
> func BenchmarkMap_AsyncExec_Different_Goroutine(b *testing.B) {
> 	var m Map
> 	for i := 0; i < b.N-1; i++ {
> 		go func() {
> 			id := uuid.New().String()
> 			m.AsyncExec(id, func() {})
> 		}()
> 	}
> }
> ```
> 
> Results:
> 
> ```
> BenchmarkMutexGroup_Lock_Same
> BenchmarkMutexGroup_Lock_Same-4                  	 3211293	       397 ns/op
> BenchmarkMap_AsyncExec_Same
> BenchmarkMap_AsyncExec_Same-4                    	  728451	      1499 ns/op
> BenchmarkMutexGroup_Lock_Different
> BenchmarkMutexGroup_Lock_Different-4             	 1978754	       609 ns/op
> BenchmarkMap_AsyncExec_Different
> BenchmarkMap_AsyncExec_Different-4               	  292650	      4535 ns/op
> BenchmarkMutexGroup_Lock_Same_Goroutine
> BenchmarkMutexGroup_Lock_Same_Goroutine-4        	  735339	      1609 ns/op
> BenchmarkMap_AsyncExec_Same_Goroutine
> BenchmarkMap_AsyncExec_Same_Goroutine-4          	 1000000	      2319 ns/op
> BenchmarkMutexGroup_Lock_Different_Goroutine
> BenchmarkMutexGroup_Lock_Different_Goroutine-4   	 1000000	      3647 ns/op
> BenchmarkMap_AsyncExec_Different_Goroutine
> BenchmarkMap_AsyncExec_Different_Goroutine-4     	 1000000	      7381 ns/op
> ```

And what were the benchmarks for the sync.Map + serialize.Executor approach?

edwarnicke commented:
gen.go:
```
package fifosync

import (
	"sync"

	"github.com/edwarnicke/serialize"
)

//go:generate go-syncmap -output serialize_map.gen.go -type Map<string,*github.com/edwarnicke/serialize.Executor>

// Map - sync.Map with key == string and value == networkservice.NetworkServiceClient
type Map sync.Map

func (m *Map) AsyncExec(id string, f func()) (ch <-chan struct{}) {
	exec,_ := m.LoadOrStore(id,&serialize.Executor{})
	return exec.AsyncExec(f)
}
```

```
BenchmarkMutexGroup_Lock_Same
BenchmarkMutexGroup_Lock_Same-16                   	 4672918	       255 ns/op
BenchmarkMap_AsyncExec_Same
BenchmarkMap_AsyncExec_Same-16                     	 2135763	       565 ns/op
BenchmarkMutexGroup_Lock_Different
BenchmarkMutexGroup_Lock_Different-16              	 4593421	       261 ns/op
BenchmarkMap_AsyncExec_Different
BenchmarkMap_AsyncExec_Different-16                	 2154997	       558 ns/op
BenchmarkMutexGroup_Lock_Same_Goroutine
BenchmarkMutexGroup_Lock_Same_Goroutine-16         	 3454956	       334 ns/op
BenchmarkMap_AsyncExec_Same_Goroutine
BenchmarkMap_AsyncExec_Same_Goroutine-16           	 1399879	       843 ns/op
BenchmarkMutexGroup_Lock_Different_Goroutine
BenchmarkMutexGroup_Lock_Different_Goroutine-16    	 1356206	       906 ns/op
BenchmarkMap_AsyncExec_Different_Goroutine
BenchmarkMap_AsyncExec_Different_Goroutine-16      	 2400020	       864 ns/op
```

edwarnicke commented:
Please also note, I'm pretty sure that removing the item from the Map at the end of the execution of the queued function is not the correct behavior, as we will likely want to keep the whatever order preserving exclusive access method we want for the life of the connection, explicitly cleaning it up on *.Close().

edwarnicke commented on pkg/tools/fifosync/mutex.go:
```diff
@@ -0,0 +1,58 @@
+// Lock acquires lock or blocks until it gets free
+func (m *Mutex) Lock() {
+	m.init.Do(func() {
+		m.ch = make(chan struct{}, 1)
```
I'm pretty sure this doesn't preserve order, as there are no guarantees whatsoever about what order blocking go routines will be awakened to enter the channel in.

edwarnicke commented on pkg/tools/fifosync/mutex.go:
```diff
@@ -0,0 +1,58 @@
+//   waiting on) 'next'
+func Mutate(locked, next *Mutex) {
+	next.init.Do(func() {
+		next.ch = make(chan struct{}, 1)
```
This doesn't preserve order.  Try running:

```go
func BenchmarkMutex_Lock_Different_Goroutine_Order(b *testing.B) {
	var g fifosync.Mutex
	check := map[int]bool{0:true}
	for i := 0; i<b.N; i++ {
		go func(i int) {
			g.Lock()
			if !check[i]{
				b.Errorf("did not find expected value in check[%d]",i)
			}
			check[i+1] = true
			g.Unlock()
		}(i)
	}
}
```

Bolodya1997 replied:
This test is not actually correct, because such code:
```go
go funcA()
go funcB()
go funcC()
```
doesn't mean that the order of execution will be `funcA()`, `funcB()`, `funcC()`.
For the order preserving testing I have added `<-time.After(time.Millisecond)` in my tests to let the goroutines start:
https://github.com/networkservicemesh/sdk/blob/a84477b8e856e9799a81745d91489f6bb5e15dfa/pkg/tools/fifosync/mutex_test.go#L36-L50

edwarnicke replied:
Change your time.Sleep to microsecond and see what happens.

What you've really demonstrated here is why Mutex semantics are not appropriate for preserving order.  The issue isn't one of implementation, its one of *semantics*.  The analogous test for AsyncExec:

```
func BenchmarkAsyncExec_Order(b *testing.B) {
	var g fifosync.Map
	check := map[int]bool{0:true}
	for i := 0; i<b.N; i++ {
		j := i
		g.AsyncExec("foo",func(){
			if !check[j]{
				b.Errorf("did not find expected value in check[%d]",i)
			}
			check[j+1] = true
		})
	}
}
```

Passes just fine, no sleep hacking needed to make it work.  I *does* guarantee order.  go routines + mutexes never do, and adding sleeps to try to make them isn't really the answer.

Bolodya1997 replied:
> What you've really demonstrated here is why Mutex semantics are not appropriate for preserving order. The issue isn't one of implementation, its one of semantics. The analogous test for AsyncExec passes just fine, no sleep hacking needed to make it work. I does guarantee order. go routines + mutexes never do, and adding sleeps to try to make them isn't really the answer.

I guess that the use case we want to cover with `MutexGroup` / `Map` is **preserving order of execution** for multiple events coming from **multiple goroutines** in some **absolute order**.
Your test is about the **execution order preserving** for the events coming from the **single goroutine** and it does assume some **absolute order** because single threaded execution is serial by nature. But it tells nothing about the **multiple goroutines** case.
Such sleep / waitgroup hacking does nothing with the **order of execution** because [in my tests the mutex is initially locked](https://github.com/networkservicemesh/sdk/blob/5df451b681f28bbc9997ce260b1a5e8d85ef59a8/pkg/tools/fifosync/mutex_group_test.go#L32). It only presumes the **order of locking**. We need it, because there is no other way to guarantee the **absolute order** of the incoming events in the **multiple goroutines** case.

> Change your time.Sleep to microsecond and see what happens.

Such thing breaks the **order of locking** so the **absolute order** of the incoming events becomes different and so it does mean that the **order of execution** must become different.

Bolodya1997 commented:
> Please also note, I'm pretty sure that removing the item from the Map at the end of the execution of the queued function is not the correct behavior, as we will likely want to keep the whatever order preserving exclusive access method we want for the life of the connection, explicitly cleaning it up on *.Close().

There can be a possible situation:
1. client fires refresh Request
2. client fires update Request
3. timer fires Close
4. `timeout` starts processing Close
5. `timeout` starts processing refresh Request
6. `timeout` ends processing Close
7. `timeout` starts processing update Request

If we delete mutex (executor) on 4, then { 5, 7 } will be serialized, but { 4, 5 } will be executed in parallel.
If we delete mutex (executor) on 6, then { 4, 5 } will be serialized, but { 5, 7 } will be executed in parallel.

So we can't simply delete mutex (executor) on Close, we can try to delete it if there is 0 usages for it at the moment - this is the reason why we still need to store `count` and the reason why do we need additional mutex (executor) to update it on lock/unlock (execution start/end).
`sync.Map` doesn't provide a way to atomically check or change such count and delete entry - so it is not the solution.

I have updated the PR with `Delete(id) bool` method that checks for the `count == 0` and deletes if so:
```
BenchmarkMutexGroup_Lock_Same
BenchmarkMutexGroup_Lock_Same-4                  	 5699420	       200 ns/op
BenchmarkMap_AsyncExec_Same
BenchmarkMap_AsyncExec_Same-4                    	  589488	      2039 ns/op
BenchmarkMutexGroup_Lock_Different
BenchmarkMutexGroup_Lock_Different-4             	 4947151	       239 ns/op
BenchmarkMap_AsyncExec_Different
BenchmarkMap_AsyncExec_Different-4               	  521427	      2460 ns/op
BenchmarkMutexGroup_Lock_Same_Goroutine
BenchmarkMutexGroup_Lock_Same_Goroutine-4        	 2486710	       472 ns/op
BenchmarkMap_AsyncExec_Same_Goroutine
BenchmarkMap_AsyncExec_Same_Goroutine-4          	  918546	      1246 ns/op
BenchmarkMutexGroup_Lock_Different_Goroutine
BenchmarkMutexGroup_Lock_Different_Goroutine-4   	 1302781	       860 ns/op
BenchmarkMap_AsyncExec_Different_Goroutine
BenchmarkMap_AsyncExec_Different_Goroutine-4     	  738832	      1590 ns/op
```
such optimization is pretty nice for the `MutexGroup`, because it reduces the count of the channel recreating, but does almost nothing for the `Map`(`Executor`) because it doesn't reduce the count of the goroutines started.

denis-tingajkin reviewed (commented):
@Bolodya1997 , @edwarnicke 

I've tried to use some FIFO mutex implementation like
```go
package fifomutex

import (
	"sync"
	"sync/atomic"
)

type FIFOMutex struct {
	head uint32
	tail uint32
	tickets map[uint32]chan struct{}
	mutex sync.Mutex
	once sync.Once
}

func (m*FIFOMutex) init()  {
	m.tickets = make(map[uint32]chan struct{})
}
func (m*FIFOMutex) Lock() {
	m.once.Do(m.init)
	ticket := atomic.AddUint32(&m.tail, 1)
	m.mutex.Lock()
	ticket = ticket - 1
	if current := atomic.LoadUint32(&m.head); current != ticket {
		sleep := make(chan struct{})
		m.tickets[ticket] = sleep
		m.mutex.Unlock()
		<-sleep
		m.mutex.Lock()
		if ticket != atomic.LoadUint32(&m.head) {
			panic("something is wrong")
		}
	}
}

func (m*FIFOMutex) Unlock() {
	m.once.Do(m.init)
	next := atomic.AddUint32(&m.head, 1)
	if sleep, ok := m.tickets[next]; ok {
		delete(m.tickets, next)
		close(sleep)
	}
	m.mutex.Unlock()
}
```

Also, I used these tests
```go
package fifomutex_test

import (
	"fmt"
	"sync"
	"testing"
	"time"
	
	"github.com/edwarnicke/serialize"
	"github.com/stretchr/testify/require"
	
	"github.com/networkservicemesh/sdk/pkg/tools/fifomutex"
)

const goroutinesCount = 1000

type asyncExecLocker struct {
	executor serialize.Executor
	locker sync.Mutex
}

func (a *asyncExecLocker) Lock()  {
	<-a.executor.AsyncExec(func() {
		// do sync ordered work
	})
}
func (a *asyncExecLocker) Unlock()  {
}


func BenchmarkFIFOMutex_LockUnlock(b *testing.B) {
	benchmarkLocker(b, new(fifomutex.FIFOMutex))
}

func BenchmarkAsyncExecutorMutex_LockUnlock(b *testing.B) {
	benchmarkLocker(b, new(asyncExecLocker))
}

func BenchmarkMutex_LockUnlock(b *testing.B) {
	benchmarkLocker(b, new(sync.Mutex))
}

func BenchmarkAsyncExecMutex_Concurrent(b *testing.B)  {
	benchmarkCrosscurrent(b, new(asyncExecLocker))
}

func BenchmarkFIFOMutex_Concurrent(b *testing.B)  {
	benchmarkCrosscurrent(b, new(fifomutex.FIFOMutex))
}

func BenchmarkMutex_Concurrent(b *testing.B)  {
	benchmarkCrosscurrent(b, new(sync.Mutex))
}


// benchmarkCrosscurrent spawns goroutinesCount goroutines
// each goroutine waits for lock
// each iteration of the test checks that all goroutines got access to "resource"
func benchmarkCrosscurrent(b *testing.B, locker sync.Locker) {
	b.StopTimer()
	b.SetParallelism(goroutinesCount)
	spawnGoroutines := func()  (waitAll func()) {
		locker.Lock()
		var wg sync.WaitGroup
		for i := 0; i < goroutinesCount; i++ {
			wg.Add(1)
			go func() {
				locker.Lock()
				locker.Unlock()
				wg.Done()
			}()
		}
		return func() {
			locker.Unlock()
			wg.Wait()
		}
	}
	
	for n := 0; n < b.N; n++ {
		b.StopTimer()
		waitAll := spawnGoroutines()
		b.StartTimer()
		waitAll()
	}
}

// benchmarkLocker just locks and unlocks passed locker N times
func benchmarkLocker(b *testing.B, locker sync.Locker) {
	for n := 0; n < b.N; n++ {
		locker.Lock()
		locker.Unlock()
	}
}

func TestFIFO(t *testing.T)  {
	require.NotPanics(t, func() {
		testFIFO(new(fifomutex.FIFOMutex))
	})
	require.Panics(t, func() {
		testFIFO(new(sync.Mutex))
	})
}


func testFIFO( l sync.Locker)  {
	l.Lock()
	var order []int
	var wg sync.WaitGroup

	for i := 0; i < goroutinesCount; i++ {
		wg.Add(1)
		go func(n int) {
			wg.Done()
			l.Lock()
			order = append(order, n)
			l.Unlock()
		}(i)
		time.Sleep(time.Millisecond*10)
	}
	wg.Wait()
	if len(order) != 0 {
		panic("order len is wrong")
	}
	l.Unlock()
	l.Lock()
	if len(order) != goroutinesCount {
		panic("order len is wrong")
	}
	for i := 0; i < goroutinesCount; i++ {
		if i != order[i] {
			println(i)
			fmt.Println(order)
			panic("order is wrong")
		}
	}
	l.Unlock()
	
}

```

I found that async.Executor could work faster than fifomutex.

## Results

Note: each test had spawned 1000 goroutines N times
```go
BenchmarkAsyncExecMutex_Concurrent-16    	    1410	    0.8 ms/op
BenchmarkFIFOMutex_Concurrent-16    	     793	    1.4 ms/op
BenchmarkMutex_Concurrent-16    	    2643	    0.4 ms/op
```

## Conclusion

So I think we should follow @edwarnicke  suggestion and just use a map of executors.

OR 

We can find proof that the channel is FIFO in load. In this case, the best solution is using FIFO mutex based on channel.
